### PR TITLE
Migrate build system from MakefIle to CMake

### PR DESCRIPTION
**Pull request by lucafuji**

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -3,28 +3,37 @@ FROM nvidia/cuda:9.1-devel-ubuntu16.04
...
+
+WORKDIR /
 
 # clone aresdb repo and set up GOPATH
 RUN mkdir -p $UBER_GITHUB_DIR
 WORKDIR $UBER_GITHUB_DIR
 RUN git clone https://github.com/uber/aresdb
 RUN ln -sf $UBER_GITHUB_DIR/aresdb $HOME/aresdb
 WORKDIR aresdb
+RUN cmake .
```
Comment (shz117) [inline docker/Dockerfile]: maybe add this to README as well.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+execute_process(COMMAND bash -c "find .  -type f \
+    \\( -iname \\*.cu -o -iname \\*.h -o -iname \\*.hpp -o -iname \\*.c \\) \
+     | grep -v \
+  -e vendor -e go-build \
+  -e build \
+  -e '.*/\\..*' \
+  -e '.*/_.*' \
+  -e '.*/mocks.*'"
+        OUTPUT_VARIABLE ALL_C_SRC)
+string(REPLACE "\n" " " ALL_C_SRC ${ALL_C_SRC})
```
Comment (shz117) [inline CMakeLists.txt]: optionally, pipe `find` output to  `| tr '\n' ' '`

Comment (shz117) [reply CMakeLists.txt]: or, change the *lang-lint.sh scripts to handle both space and new line separated input

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        COMMAND ./scripts/clang-lint.sh ${ALL_C_SRC}
+        COMMAND ./scripts/golang-lint.sh ${ALL_GO_SRC}
+        VERBATIM)
+
+add_custom_target(lint
+        COMMAND ./scripts/clang-lint.sh ${CHANGED_C_SRC}
+        COMMAND ./scripts/golang-lint.sh ${CHANGED_GO_SRC}
+        VERBATIM)
+
+add_custom_target(golang-dep DEPENDS vendor/glide.updated)
```
Comment (shz117) [inline CMakeLists.txt]: is it possible merge this two lines with `add_custom_target` with its own commands?

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        query/transform.hpp
+        query/utils.cu
+        query/utils.hpp)
+
+
+################################
+# GTest
+################################
+set(GTEST_ROOT $ENV{GTEST_ROOT})
+link_directories(${GTEST_ROOT})
```
Comment (shz117) [inline CMakeLists.txt]: possible to merge to `ink_directories($ENV{GTEST_ROOT})` ?

File: CMakeLists.txt (hunk 3 above)
Comment (lucafuji) [reply CMakeLists.txt]: no, have to be one target and one commands and the command to generate the fileshttps://stackoverflow.com/questions/7862300/how-to-make-cmake-target-executed-whether-specified-file-was-changed

File: CMakeLists.txt (hunk 2 above)
Comment (lucafuji) [reply CMakeLists.txt]: we need those variable.e.g ALL_GO_SRC as dep files

File: CMakeLists.txt (hunk 4 above)
Comment (lucafuji) [reply CMakeLists.txt]: we reference GTEST_ROOT multiple times later, see target_include_directories(all_unittest PRIVATE ${GTEST_ROOT}/include ${gtest_SOURCE_DIR})

File: docker/Dockerfile (hunk 1 above)
Comment (lucafuji) [reply docker/Dockerfile]: added

File: CMakeLists.txt (hunk 3 above)
Comment (shz117) [reply CMakeLists.txt]: I see, good to know

**Pull request merged**